src/parse.rs

```diff
@@ -1028,14 +1028,18 @@ fn append_text<'input, 'temp>(
             if let NodeKind::Text(ref mut prev_text) = node.kind {
                 let text = text.as_str();
                 match prev_text {
-                    Cow::Borrowed(s) => {
+                    SharedString::Borrowed(s) => {
                         let mut concat_text = String::with_capacity(s.len() + text.len());
                         concat_text.push_str(s);
                         concat_text.push_str(text);
-                        *prev_text = Cow::Owned(concat_text);
+                        *prev_text = SharedString::Owned(Arc::from(concat_text));
                     }
-                    Cow::Owned(ref mut s) => {
-                        s.push_str(text);
+                    SharedString::Owned(s) => {
+                        // TODO: find a way to not reallocate the string.
```

Comment:
I am really weary about this part, because with `Arc<str>` it means dropping excess capacity when concatenating the strings. I wonder if a type like

```rust
pub enum Text<'input> {
  Borrowed(&'input str),
  Owned(String),
  Shared(Arc<str>),
}
```

would be nicer because of two reasons:
* We could use `Text::Owned(..)` while extending text nodes like here.
* Going from `Text::Owned(..)` to `Text::Shared(..)` could be single optional post-processing step when we are sure to never touch a text node again.

If that post-processing step to go from `Text::Owned(..)` to `Text::Shared(..)` would be controlled by `ParseOptions`, this would also mean that users who do not want shared text could avoid the reference counting completely. (We already pay for the enum tag with `Cow`. Sadly, we cannot make `Arc` versus `Rc` decidable at runtime.)

Reply:
Actually, the sharing could be fully lazy without an option in `ParseOptions`: We always produce `Text::Owned(..)` during parse and your `shared_value` becomes:

```
pub fn shared_value(&mut self) -> Arc<str> {
  match self.data.value {
    Text::Borrowed(text) => {
      let text = Arc::from(text),
      self.data.value = Text::Shared(text.clone());
      text      
    },
    Text::Owned(text) => {
      let text = Arc::from(text);
      self.data.value = Text::Shared(text.clone());
      text
    }
    Text::Shared(text) => text.clone(),
  }
}
```

Reply:
That way, we do not pay anything we are not already paying and `shared_value` et al. are also not paying anything they do not in the up-front approach expect for the `&mut self` requirement.

Reply:
So this "wonderful" idea of on-demand sharing does not work as `Attribute` has not access to `&mut AttributeData`. The `ParserOption` route might still be viable though. I am currently looking into that...

Reply:
I tried to resolve this TODO but I do not think it can be done safely as going from `String` to `Arc<str>` will copy the text into a new allocation where it is prepend with `Arc`'s header. It is then not possible to get a mutable variant like `String` back for this allocation without doing something about that header, e.g. dropping it thereby requiring a new allocation. (This also means that in the case of the previous node being borrowed, we also do one more copy of the concatenated text. It would surprise me if benchmarks cannot detect this amount of additional copying?)

Reply:
So since this is copying anyway, we only need to amortize the extra allocations, e.g.

```diff
diff --git a/src/parse.rs b/src/parse.rs
index 36cb3ab..a4e9475 100644
--- a/src/parse.rs
+++ b/src/parse.rs
@@ -370,6 +370,7 @@ struct ParserData<'input> {
     parent_prefixes: Vec<&'input str>,
     entities: Vec<Entity<'input>>,
     buffer: TextBuffer,
+    concat_buffer: String,
     after_text: bool,
 }
 
@@ -481,6 +482,7 @@ fn parse(text: &str, opt: ParsingOptions) -> Result<Document, Error> {
         awaiting_subtree: Vec::new(),
         parent_prefixes: Vec::new(),
         buffer: TextBuffer::new(),
+        concat_buffer: String::new(),
         after_text: false,
     };
 
@@ -585,6 +587,7 @@ fn process_tokens<'input>(
                     parent_id,
                     span.start(),
                     pd.opt.nodes_limit,
+                    &mut pd.concat_buffer,
                     pd.after_text,
                     doc,
                     &mut pd.awaiting_subtree,
@@ -910,6 +913,7 @@ fn process_text<'input>(
             parent_id,
             text.start(),
             pd.opt.nodes_limit,
+            &mut pd.concat_buffer,
             pd.after_text,
             doc,
             &mut pd.awaiting_subtree,
@@ -953,6 +957,7 @@ fn process_text<'input>(
                         parent_id,
                         text.start(),
                         pd.opt.nodes_limit,
+                        &mut pd.concat_buffer,
                         pd.after_text,
                         doc,
                         &mut pd.awaiting_subtree,
@@ -980,6 +985,7 @@ fn process_text<'input>(
             parent_id,
             text.start(),
             pd.opt.nodes_limit,
+            &mut pd.concat_buffer,
             pd.after_text,
             doc,
             &mut pd.awaiting_subtree,
@@ -1012,12 +1018,13 @@ impl<'input, 'temp> BorrowedText<'input, 'temp> {
     }
 }
 
-#[allow(clippy::needless_lifetimes)]
+#[allow(clippy::needless_lifetimes, clippy::too_many_arguments)]
 fn append_text<'input, 'temp>(
     text: BorrowedText<'input, 'temp>,
     parent_id: NodeId,
     pos: usize,
     nodes_limit: u32,
+    concat_buffer: &mut String,
     after_text: bool,
     doc: &mut Document<'input>,
     awaiting_subtree: &mut Vec<NodeId>,
@@ -1027,21 +1034,12 @@ fn append_text<'input, 'temp>(
         if let Some(node) = doc.nodes.last_mut() {
             if let NodeKind::Text(ref mut prev_text) = node.kind {
                 let text = text.as_str();
-                match prev_text {
-                    SharedString::Borrowed(s) => {
-                        let mut concat_text = String::with_capacity(s.len() + text.len());
-                        concat_text.push_str(s);
-                        concat_text.push_str(text);
-                        *prev_text = SharedString::Owned(Arc::from(concat_text));
-                    }
-                    SharedString::Owned(s) => {
-                        // TODO: find a way to not reallocate the string.
-                        let mut concat_text = String::with_capacity(s.len() + text.len());
-                        concat_text.push_str(s);
-                        concat_text.push_str(text);
-                        *prev_text = SharedString::Owned(Arc::from(concat_text));
-                    }
-                }
+                let s = prev_text.as_str();
+                concat_buffer.clear();
+                concat_buffer.reserve(s.len() + text.len());
+                concat_buffer.push_str(s);
+                concat_buffer.push_str(text);
+                *prev_text = SharedString::Owned(Arc::from(concat_buffer.as_str()));
             }
         }
     } else {
```

Reply:
I think this can be done separately. Afaik, the second branch is executed only when we have sequential CDATA, which is extremely rare.

Reply:
Agreed. I volunteer for a follow-up PR. :-) (The buffer would also amortize the allocation in the first branch.)